docs/HTTP_V2_SPEC.md

```diff
@@ -1,106 +1,236 @@
 # The sqld HTTP API v2 specification ("Hrana over HTTP")
 
-Version 2 of the HTTP API ("Hrana over HTTP") extends the HTTP API with features
-introduced in Hrana 2.
+Version 2 of the HTTP API ("Hrana over HTTP") exposes stateful streams from
+Hrana over HTTP. It provides functionality equivalent to Hrana and it is useful
+for environments with missing or incomplete support for WebSockets.
+
+This version deprecates version 1 of the HTTP API. Both clients and servers
+should move to version 2 as soon as possible.
 
 ## Overview
 
-This HTTP API uses data structures and semantics from the Hrana 2 protocol.
-Endpoints in the HTTP API correspond to requests in Hrana. Each request is
-executed as if a fresh Hrana stream was opened for the request.
+The HTTP API uses data structures and semantics from the Hrana 2 protocol.
 
-All request and response bodies are encoded in JSON, with content type
-`application/json`.
+Individual requests on the same stream are tied together by the use of a baton.
+The server returns a baton in every response to a request on the stream, and the
+client then needs to include the baton in the subsequent request. The client
+must serialize the requests: it must wait for a response to the previous request
+before sending next request.
 
-## Version check
+The server can also optionally specify a different URL that the client should
+use for the requests on the stream. This can be used to ensure that stream
+requests are "sticky" and reach the same server.
 
-```
+The server will close streams after a short period of inactivity, to make sure
+that abandoned streams don't accumulate on the server.
```

Comment:
since we should be closing the stream if the client closes it (TCP), I'd highly recommend we allow the client to specify the `timeout=seconds` in the url param. Of course, default to some reasonable timeout, but let users do long-lived sessions whenever needed.

In addition to that, if we want to be resilient to quick network failures, we could allow the `session` in the control-url returned in my example below to be given as another url param, ex: `reconnect=session`

Then connections that are not explicitly closed via control_url (ie: TCP disconnect/timeout detected by the server) will be put in a "disposal queue" instead of closed immediately.

If the new stream is created with the `reconnect=session` as parameter, then it will see if it's still pending disposal and will bring it to use again.

New control requests should also remove from the disposal queue and bring it to life again instead of failing (suppose the stream fails tcp/networking, but a controls comes in, then the stream reconnects).

Whenever the reconnection happens, the pending responses are flushed to the stream.

This reconnection timeout should also be configurable, or at least enabled/disabled... in some scenarios this may be undesired.

NOTE: to reliably check the client, the server should be sending some `: ping\n\n` messages to the client, they should be ignored (https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format), but this should lead the tcp to fail if the client is gone.

Reply:
The only point of keeping a long-living TCP connection would be to close the stream immediately when the connection is broken. Introducing the disposal queue would nullify this advantage.

Also, environments that can establish a reliable long-living TCP connection can just use WebSockets. The HTTP API is designed for environments where we can only perform simple HTTP requests.

Reply:
I don't understand the first sentence :)

Long polling is more complex and more fragile than "normal" HTTP requests and we can implement everything that we need without long polling, so why use it?

Reply:
> I don't understand the first sentence :)

the environment (aws lambda, azure functions and whatever runs on top of them, like vercel) cannot detect if the HTTP request is meant to be used in a long polling or not. All they can (and do) is impose some timeout on the request.


> Long polling is more complex and more fragile than "normal" HTTP requests and we can implement everything that we need without long polling, so why use it?

because you can detect failures and misbehavior. I also don't think it's complex, unless we're imagining different things